### DISCUSSION TITLE
Implementation in Django

### DISCUSSION THREAD

**Discussion by Marco-Gtto** [miguelgrinberg/python-socketio#744] [Q&A]
Hi @miguelgrinberg,
I am developing a project with Django, in which I have recently started integrating socket.io.
I followed the documentation and your specific examples about it, getting to this point:

views.py
```
from django.shortcuts import render
import socketio
from .models import User

sio = socketio.Server(async_mode='eventlet', logger=True)

@sio.event
def connect(sid, environ):
    print('connect ', sid)


@sio.event
def disconnect(sid):
    print('disconnect ', sid)


@sio.on('user_update')
def user_update(sid, data):
    pass


@sio.on('user_sanction')
def user_sanction(sid, data):
    pass
```

wsgi.py 
```
import os
import eventlet
import socketio
import eventlet
import eventlet.wsgi
from django.core.wsgi import get_wsgi_application

os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'discordwebsite.settings')

application = get_wsgi_application()
sio = socketio.Server()
application = socketio.WSGIApp(sio, application)

eventlet.wsgi.server(eventlet.listen(('', 5000)), application)

```

In my project I created an app that will only take care of updating the database in relation to the data received via socket. I have currently started writing code for this in views.py, although I would like to create a dedicated file for this purpose. Essentially the problem is that in the console I read the following strings:

```
(8180) wsgi starting up on http://0.0.0.0:5000
(8180) accepted ('127.0.0.1', 50224)
```

however the event handlers in views.py seem not to exist, as after connection I don't receive any message in the console.
I'm not actually able to figure out if I need to invoke something somewhere or add more in wsgi to get the event handlers in question to work.
I hope for some advice that can help me continue.

Thanks in advance
Marco

**Comment by miguelgrinberg** [comment]
Are you importing views.py somewhere in your application? The file needs to be imported for the handlers to be registered with the server.

**Comment by Marco-Gtto** [comment]
So, the situation is the following:
 you were right about importing views, I corrected this problem, also to make everything work I had to import the "sio" instance directly from views into wsgi. So the socket.io server is working, however django's is not, because if I open the address 127.0.0.1:8000 on the browser I get a connection refused error. If I remove the last line from wsgi that you saw in my original message, the server starts working again. I guess the eventlet server conflicts with the django server, even though I have specified different ports for the development server and the socket server. Is there any way around this problem?

Another question I would like to ask is about the example you uploaded here on Github. There is no eventlet listeners and / or similar in your wsgi file. How does the socket server listen for events? To be safe, I copied your example comma by comma and the socket server never started

I look forward to hearing from you
Thank you
Marco

**Comment by Marco-Gtto** [comment]
Ok, after a few minutes of banging my head on the table after my last answer, I thought: Why not try setting the same port as the development server in the eventlet server? and it worked! I was simply convinced from the beginning that the listening port of the eventlet server should be different from that of the development server, but evidently it wasn't.
So the only question I would be interested in answering is the one about the missing eventlet server in the github example.

**Comment by miguelgrinberg** [comment]
@Marco-Gtto I think part of your misunderstanding comes from the fact that you are thinking there are two servers. There's only one server and in terms of WSGI there is only one WSGI application.

In your `wsgi.py` file you are hijacking the control Django has over the web server, and instead launching your own eventlet web server, which replaces the Django web server. This is not a good idea, in my opinion, because the wsgi.py file is supposed to export the `application`, not launch a web server of its own and then block to listen for requests.

The approach I've taken in the Django example in this repository is cleaner, in my opinion. In the wsgi.py file all I'm doing is creating the combined WSGI application. This would allow you to run the application with any web server. If you wanted to run the application under eventlet, then you could write a `run.py` or similar file in the root directory of your project and import the application from wsgi.py:

```python
import eventlet
import eventlet.wsgi
from myproject.wsgi import application

eventlet.wsgi.server(eventlet.listen(('', 5000)), application)
```

But if you prefer to use Gunicorn, you can do that, without having to modify your source code:

```bash
gunicorn -b :5000 -k eventlet -w 1 myproject.wsgi
```

The answer to your question regarding the "missing" eventlet server is that the application is an application, not a web server. Which web server you run it on is not something that needs to be expressed inside the application. To run my example under eventlet you can use any of the two approaches I showed you above. Also for convenience, I have added a customized `runserver` command, so yet another way to start the project would be to use `python manage.py runserver`.
